### PR TITLE
Breaking upgrade docs (0.4)

### LINKED ISSUE #1048: Upgrade to 0.4.0 [documentation]
We need to work out communications to help upgrade to version 0.4.0. This could be a combination of release notes and docs.

There are [general upgrade docs](https://docs.qhub.dev/en/latest/source/admin_guide/upgrade.html) ready to be released but these are not specific to 0.4.0 which may need special consideration especially in light of #1040.

# My suggested process

Assumes #1047 and #957 have been merged.

## Backup and move existing URL

Using the old version of qhub:

1. Perform manual backups (Complication: after release the docs will include details about Keycloak, but the only thing relevant will just be Zipping up the NFS files).

2. Rename the existing domain

This will allow the existing cluster to remain alive in case it is needed, but the idea would be not to have it in use from now on.

In the qhub-config.yaml for example:
```
project_name: myqhub
domain: qhub.myproj.com
```
could become:
```
project_name: myqhub
domain: qhub-old.myproj.com
```

Run `qhub deploy` using the existing (old) version of qhub.

## Upgrade the config file

Now we need to go back to the original qhub-config.yaml file which contained the original domain.

Run the [general upgrade docs](https://docs.qhub.dev/en/latest/source/admin_guide/upgrade.html) process up to the point before redeployment.

**Actually, I've noticed that steps 5 and 6 are specific to the 0.4.0 upgrade, so should probably be marked as such.** They could be moved to 0.4.0-specific documentation, but of course they may also be required if someone upgrades directly from 0.3.14 -> 0.4.1.

You should end up with a new `qhub-config.yaml` file and a `qhub-users-import.json`.

## Rename the project

Need to rename the project to avoid clashes with the existing (old) cluster. The domain should remain as the preferred main one that was always in use previously.

For example:
```
project_name: myqhub
domain: qhub.myproj.com
```
could become:
```
project_name: myqhubnew
domain: qhub.myproj.com
```

## Deploy and restore backups

When attempting to redeploy, it should first fail due to `prevent_deploy` flag needing to be removed for safety.

After deployment, carry out steps 5 and 6 of the general upgrade docs (OAuth URL updates if relevant, plus inport Keycloak users).

Change the Keycloak root password at /auth URL - don't update the qhub-config.yaml file to reflect this.

Restore NFS from backup - docs needed...

Also need to chown all uids immediately after restore while the 'restore pod' is mounted:

```
chown -R 1000:100 /mnt/home
```

### PR DESCRIPTION
**Pull request by danlester**
Closes #1048

Docs about 0.4-specific upgrade.

## Changes:

- Added a `breaking-upgrade.md` doc that contains more detail upgrade steps when it entails a redeploy (compared to regular simple `upgrade.md`)
- Updated `backup.md` to reflect new NFS file structure including mapping during 0.4 upgrade
- Brought in specific commands required to upload/download backups to S3 using only curl (in `awss3curl.md`)

## Types of changes

What types of changes does your code introduce?

_Put an `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features to not work as expected)
- [x] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

### Requires testing

- [x] Yes
- [ ] No

### In case you checked yes, did you write tests?

- [ ] Yes
- [x] No

### CODE REVIEW

File: docs/source/admin_guide/backup.md [hunk 1]
```diff
@@ -145,6 +147,28 @@ tar -xvf 2021-04-23.tar
 
 The file permissions for the default tar is same as the original files.
 
+> **Important: If upgrading from 0.3.14 or earlier to 0.4 or later**
+>
+> QHub v0.4: If restoring your NFS as part of the upgrade you must also run some extra commands, immediately after extracting from the tar file.
+>
+> Because previous versions contained the `shared` folder within `home`, but they are now both at the same level, you must run:
+> ```
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:155]:
```suggestion
> ```bash
```

File: docs/source/admin_guide/backup.md [hunk 2]
```diff
@@ -145,6 +147,28 @@ tar -xvf 2021-04-23.tar
...
+> cd /data
+> cp -r home/shared/* shared/
+> rm -rf home/shared/
+> ```
+> And then:
+> ```
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:161]:
```suggestion
> ```bash
```

File: docs/source/admin_guide/backup.md [hunk 3]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
 Similar instructions, but use Digital Ocean spaces. This guide explains installation of the command-line tool:
 https://www.digitalocean.com/community/tutorials/how-to-migrate-from-amazon-s3-to-digitalocean-spaces-with-rclone
 
+
+## JupyterHub Database
+
+The JupyterHub database will mostly be recreated when you start a new cluster anyway, but should be backed up to save Dashboard configurations.
+
+You can do something very similar to the NFS backup, above - this time you need to back up a file on the PersistentVolume `hub-db-dir`. First, you might think you can just make a new `pod.yaml` file, this time specifying `claimName: "hub-db-dir"` instead of `claimName: "nfs-mount-dev-share"`. However, `hub-db-dir` is 'Read Write Once' - the 'Once' meaning it can only be mounted to one pod at a time, but the JupyterHub pod will already have this mounted!
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:194]: I got really lost following this section - could probably be turned into a list of numbered items specifying the steps

File: docs/source/admin_guide/backup.md [hunk 4]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+
+So instead of mounting to a new 'debugger pod' we have to access the JupyterHub pod directly and see what we can do from there.
+
+Look up the JupyterHub pod:
+```
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:199]:
```suggestion
```bash
```

File: docs/source/admin_guide/backup.md [hunk 5]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+kubectl get pods -n dev
+```
+
+It will be something like `hub-765c9488d6-8z4nj`.
+
+Get a shell into that pod:
+
+```
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:207]:
```suggestion
```bash
```

File: docs/source/admin_guide/backup.md [hunk 6]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+kubectl exec -n dev --stdin --tty hub-765c9488d6-8z4nj -- /bin/bash
+```
+
+There is no need to TAR anything up since the only file required to be backed up is `/srv/jupyterhub/jupyterhub.sqlite`.
+
+### Backing up JupyterHub DB
+
+So we just need to upload the file to S3. You might want to install the AWS CLI tool as we did before, but unfortunately the Hub container is quite locked down and it isn't straightforward to install that... You might need to upload to S3 using curl directly:
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:215]: This sentence is duplicated from somewhere else above - but I did not see any actual steps on how to install said CLI

File: docs/source/admin_guide/backup.md [hunk 7]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+
+As [explained here](./awss3curl.md).
+
+### Restoring JupyterHub DB
+
+Simply overwrite the same file `/srv/jupyterhub/jupyterhub.sqlite` based on the version backed up to S3.
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:221]:
```suggestion
You will need to overwrite the file `/srv/jupyterhub/jupyterhub.sqlite` based on the version backed up to S3.
```

File: docs/source/admin_guide/backup.md [hunk 8]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+
+You should restart the pod:
+```
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:224]:
```suggestion
```bash
```

File: docs/source/admin_guide/backup.md [hunk 9]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+kubectl delete -n dev pod hub-765c9488d6-8z4nj
+```
+
+As for uploads, [you may need to download using curl](./awss3curl.md).
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:228]:
```suggestion
As for uploads, [you may need to use curl to download items from an AWS S3 bucket](./awss3curl.md).
```

File: docs/source/admin_guide/backup.md [hunk 10]
```diff
@@ -162,6 +186,47 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+
 ## Keycloak user/group database
 
 QHub provides a simple script to export the important user/group database. Your new QHub cluster will recreate a lot of Keycloak config (including new Keycloak clients which will have new secrets), so only the high-level Group and User info is exported.
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:232]: is this accurate? will it create a lot or all of?

File: docs/source/admin_guide/backup.md [hunk 11]
```diff
@@ -193,28 +258,3 @@ To re-import your users and groups, [login to the /auth/ URL](../installation/lo
 Under 'Manage' on the left-hand side, click 'Import'. Locate the `exported-keycloak.json` file and select it. Then click the 'Import' button.
 
 All users and groups should now be present in Keycloak. Note that password will not have been restored so may need to be reset.
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:260]:
```suggestion
All users and groups should now be present in Keycloak. Note that the passwords will not have been restored so you may need to be reset them after this step.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
+# Upgrade - Redeployment for Breaking Changes
+
+For versions that are known to require redeployment when upgrading, the steps for upgrade include:
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:3]:
```suggestion
For versions that are known to require redeployment when upgrading,  follow the next steps to upgrade QHub:
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+- Back up user data by following [this guide](./backup.md).
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:5]: It is recommended to use numbered lists for steps that must be performed in a given order 
```suggestion
1. Back up user data by following [the backup guide](./backup.md) in the documentation.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+- Change existing cluster to a different URL (e.g. `qhub-old.mycompany.com`) so it is hidden.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:6]:
```suggestion
- Change the existing cluster to a different URL (e.g. `qhub-old.mycompany.com`) so it is hidden.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+- Run `qhub upgrade` to make recommended modifications to the `qhub-config.yaml` file.
+- Deploy a new QHub with the desired version (to your original preferred URL e.g. `qhub.mycompany.com` but a new project_name to avoid resource name clashes).
+- Restore user data to the new cluster.
+- Destroy old cluster when happy.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:10]:
```suggestion
- Once you have confirmed the backup and upgrade have been successfully completed, you can destroy the old cluster.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+Please always check release notes for more details - and in all cases, backup your data before upgrading.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:12]: Can we add a link to where the release notes live?

File: docs/source/admin_guide/breaking-upgrade.md [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+**The rest of this guide assumes you are upgrading from version v0.3.14 to v0.4.**
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:14]: Should this be a note instead of a bold styling? we want to make this as visible as possible

File: docs/source/admin_guide/breaking-upgrade.md [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+You may be deploying QHub based on a local configuration file, or you may be using CI/CD workflows in GitHub or GitLab. Either way, you will need to locate a copy of your `qhub-config.yaml` configuration file to upgrade it (and commit back to your git repo in the CI/CD case).
+
+For CI/CD deployments, you will need to `git clone <repo URL>` into a folder on your local machine if you haven't done so already.
+
+## 1. Backup existing data
+
+Perform manual backups of the NFS data and JupyterHub database (ignore the section about Keycloak data since that will not exist in your v0.3.14 cluster).
+
+The [Manual Backup guide is here](./backup.md).
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:24]:
```suggestion
For more details please refer to the  [Manual Backup guide](./backup.md).
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+
+## 2. Rename existing QHub URL
+
+This will allow the existing cluster to remain alive in case it is needed, but the idea would be not to have it in use from now on.
+
+In the qhub-config.yaml for example:
+
+```
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:32]:
```suggestion
```yaml
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+project_name: myqhub
+domain: qhub.myproj.com
+```
+
+could become:
+
+```
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:39]:
```suggestion
```yaml
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+project_name: myqhub
+domain: qhub-old.myproj.com
+```
+
+Run qhub deploy using the existing (old) version of qhub.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:44]:
```suggestion
Run the command `qhub deploy` using the existing (old) version of QHub.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+
+## 3. Upgrade the `qhub` command package
+
+To install (or upgrade) your pip installation of the Python package used to manage QHub:
+
+```shell
+pip install --upgrade qhub==0.4.0
+```
+
+## 4. Upgrade `qhub-config.yaml` file
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:54]:
```suggestion
## 4. Upgrade the `qhub-config.yaml` file
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+
+Now we need to go back to the original `qhub-config.yaml` file which contained the original domain.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:56]: There are some inconsistencies between the usage of we and you for instructions

File: docs/source/admin_guide/backup.md [hunk 24]
```diff
@@ -145,6 +147,28 @@ tar -xvf 2021-04-23.tar
...
+> This is because all users have the same uid in QHub v0.4 onward.
+>
+
+
+
+
+
 ### Google cloud provider
 
 To use the Google Cloud provider, install [gsutil](https://cloud.google.com/storage/docs/gsutil_install). The instructions are same as the preceding steps. Additionally, use these commands for copy/download of the backup:
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:174]:
```suggestion
To use the Google Cloud provider, install the [gsutil](https://cloud.google.com/storage/docs/gsutil_install) CLI. The instructions are the same as the preceding steps. Additionally, use these commands for copy/download of the backup:
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+
+You will now have a `qhub-config.yaml` file that you can deploy.
+
+```shell
+qhub deploy -c qhub-config.yaml
+```
+
+At this point you will see an error message saying that deployment is prevented due to the `prevent_deploy` setting in your YAML file. This is a safeguard to ensure that you only proceed if you are aware of possible breaking changes in the current upgrade.
+
+Because you are reading this guide, you already know that you needed to backup your data - see above again, if not. So you can edit your `qhub-config.yaml`, find the line reading `prevent_deploy: true` and remove it.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:100]: This sentence reads rather unfriendly...

```suggestion
Make sure to **backup your data** as described in the [backup section of the documentation]<add link here>.
Only after backing up your data proceed to remove the `prevent_deploy: true` line in the `qhub-config.yaml` file.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+
+Attempt the `qhub deploy -c qhub-config.yaml` command again and it should get further this time.
+
+## 7. CI/CD: render and commit to git
+
+For CI/CD (GitHub/GitLab) workflows, then as well as generating the updated `qhub-config.yaml` files as above, you will also need to regenerate the workflow files based on the latest `qhub` version's templates.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:106]:
```suggestion
For CI/CD (GitHub/GitLab) workflows, as well as generating the updated `qhub-config.yaml` files as above, you will also need to regenerate the workflow files based on the latest `qhub` version's templates.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+
+(Note that `qhub deploy` would have performed this render step too, but will also immediately redeploy your QHub. Run the render command alone if you are now working separately in your repo and don't want to redeploy.)
+
+Commit all the files (`qhub-config.yaml` and GitHub/GitLab workflow files) back to the remote repo. All files need to be committed together in the same commit.
+
+## 8. Update OAuth callback URL for Auth0 or GitHub
+
+If your QHub deployment relies on Auth0 or GitHub for authentication, please update the OAuth callback URL.
+
+<details><summary>Click for more detailed instructions </summary>
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:122]:
```suggestion
<details><summary>Expand this section for more detailed instructions </summary>
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+
+1. Navigate to the your Auth0 tenacy homepage and from there select "Applications".
+
+2. Select the "Regular Web Application" with the name of your deployment.
+
+3. Under the "Application URIs" section, paste the new OAuth callback URL in the "Allowed Callback URLs" text block.
+- The URL will take the shape:
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:129]:
```suggestion
- The URL will look something like this:
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+    https://{your-qhub-domain}/auth/realms/qhub/broker/auth0/endpoint
+    ```
+
+Updating the GitHub Application instead is very similar, but the OAuth callback will be at `https://{your-qhub-domain}/auth/realms/qhub/broker/github/endpoint`
+
+</details>
+
+## 9. Restore from Backups
+
+Next, return to the [Manual Backups documentation](./backup.md) to perform:
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:140]:
```suggestion
Next, you will need to perform the following steps to restore from a previously generated backup as described in the [Manual Backups documentation](./backup.md):
```

File: docs/source/admin_guide/upgrade.md [hunk 30]
```diff
@@ -1,5 +1,14 @@
 # Upgrade
 
+This is a guide to the general upgrade of QHub to a new version.
+
+You should always backup your data before upgrading.
+
+> Note that for some releases (e.g. to v0.4), the cluster cannot be upgraded in-situ so you must perform a redeployment (backup the old cluster, redeploy a new upgraded cluster and then restore your data).
+>
+> To perform a redeployment upgrade see the [more detailed documentation here](./breaking-upgrade.md).
```
Comment (trallard) [inline docs/source/admin_guide/upgrade.md:9]:
```suggestion
> To perform a redeployment upgrade see the [more braking upgrade documentation](./breaking-upgrade.md).
```

**Review by trallard — CHANGES_REQUESTED** [review]

**Comment by trallard** [comment]
I made a bunch of comments and suggestions in the documentation - in brief I found quite a few sections quite hard to follow and these docs should be almost self-contained for folks to follow. From experience, many people will just skim and look for commands almost to copy and paste so we need to provide the needed steps and commands in a clear way

File: docs/source/admin_guide/backup.md (hunk 3 above)
Comment (danlester) [reply docs/source/admin_guide/backup.md:194]: I'll split this up a bit, but in fact it is not a list of steps that you should carry out. It explains why the same approach as used for NFS won't quite work this time.

File: docs/source/admin_guide/backup.md (hunk 10 above)
Comment (danlester) [reply docs/source/admin_guide/backup.md:232]: Yes, the point is that there is no need to export everything since some of the config will be created again on a new cluster, automatically. And we should use a lot of the new config - it's just things like user and group info that we want to re-import from any backup.

I don't think it's worth expanding on it further - most people will follow blindly anyway, but we just want to add a caveat to explain that we aren't actually backing up the full database, in case that's what users might be expecting.

File: docs/source/admin_guide/upgrade.md [hunk 31]
```diff
@@ -1,5 +1,14 @@
 # Upgrade
```
[the remark this answers is not in the document]
Comment (danlester) [reply docs/source/admin_guide/upgrade.md:1]: Yes, many steps are similar.

**Comment by danlester** [comment]
Thank you for your review @trallard - it was very helpful. I've implemented most suggestions.

File: docs/source/admin_guide/awss3curl.md [hunk 32]
```diff
@@ -0,0 +1,71 @@
+# Using Curl to access AWS S3
+
+In some situations, users may wish to upload content to S3 or download content from an S3 bucket. For example, when attempting [manual backups of QHub's data](./backup.md).
```
Comment (trallard) [inline docs/source/admin_guide/awss3curl.md:3]: Do we have standards re line-lenght? I generally suggest having something ~220/180 for md documents and sentence-split to help with PR review.

File: docs/source/admin_guide/awss3curl.md [hunk 33]
```diff
@@ -0,0 +1,71 @@
...
+
+Where possible, it is easiest to use AWS's command line tools to do so, but in some situations - for example, to backup the JupyterHub SQLite database - it may be difficult to install AWS' CLI tools due to being in a restricted container environment. In that situation, it is possible to fall back on AWS' basic REST API and use HTTPS requests directly instead. (Ultimately, the AWS CLI is simply a wrapper around those REST APIs.)
```
Comment (trallard) [inline docs/source/admin_guide/awss3curl.md:5]:
```suggestion
In many cases, the most straightforward way to access AWS S3 buckets is by installing and using AWS's command-line tool. But in some situations - for example, to back up the JupyterHub SQLite database - it may be difficult to install AWS' CLI tools due to being in a restricted container environment. In that situation, it is possible to fall back on AWS' basic REST API and use HTTPS requests directly instead. (Ultimately, the AWS CLI is simply a wrapper around those REST APIs.)
```

File: docs/source/admin_guide/awss3curl.md [hunk 34]
```diff
@@ -0,0 +1,71 @@
...
+
+This document describes simple curl commands to interface with S3 directly, specifically in the case of uploading a [backup](./backup.md) of JupyterHub's SQLite database from a restricted pod to S3 (or restoring it from a backup from S3).
```
Comment (trallard) [inline docs/source/admin_guide/awss3curl.md:7]:
```{suggestion}
This document describes how to use `curl` commands to interface with S3 directly, specifically in the case of [uploading a backup of JupyterHub's SQLite database](./backup.md) from a restricted pod to S3
(or restoring it from a backup from S3).
```

File: docs/source/admin_guide/backup.md [hunk 35]
```diff
@@ -41,7 +41,7 @@ aws eks update-kubeconfig  --region us-west-2 --name <relevant-name>
 
 ### Pod deployment
 
-With kubectl configured, the user now needs to deploy the pod that allows the user to access the cluster files. First save the follow pod specification to a file named `pod.yaml`
+With kubectl configured, the user now needs to deploy the pod that allows the user to access the cluster files. First save the following pod specification to a file named `pod.yaml`
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:44]:
```{suggestion}
With kubectl configured,  the next step will be to deploy the pod that allows the user to access the cluster files. First save the following pod specification to a file named `pod.yaml`
```

File: docs/source/admin_guide/backup.md [hunk 36]
```diff
@@ -63,6 +63,8 @@ spec:
           name: volume-to-debug-ubuntu
 ```
 
+> Note in QHub versions before v0.4 replace `claimName: "jupyterhub-dev-share"` with `claimName: "nfs-mount-dev-share"`
+
 Run:
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:68]:
```{suggestion}
Once the file `pod.yml` has been created, run the following command:
```

File: docs/source/admin_guide/backup.md [hunk 37]
```diff
@@ -63,6 +63,8 @@ spec:
...
 
 ```shell
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:70]: Suggestion replace `dev` with <your-namespace> from here on

Comment (trallard) [reply docs/source/admin_guide/backup.md:70]: Then you can remove some bits referring to the change from `dev` to other namespaces

File: docs/source/admin_guide/backup.md [hunk 38]
```diff
@@ -145,9 +147,27 @@ tar -xvf 2021-04-23.tar
 
 The file permissions for the default tar is same as the original files.
 
+> **Important: If upgrading from 0.3.14 or earlier to 0.4 or later**
+>
+> QHub v0.4: If restoring your NFS as part of the upgrade you must also run some extra commands, immediately after extracting from the tar file.
+>
+> Because previous versions contained the `shared` folder within `home`, but they are now both at the same level, you must run:
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:154]:
```{suggestion}
> Previous versions contained the `shared` folder within `home`. From `0.4.0` both `shared` and `home` directories are at the same level with respect to the QHub filesystem:

```

File: docs/source/admin_guide/backup.md [hunk 39]
```diff
@@ -145,9 +147,27 @@ tar -xvf 2021-04-23.tar
...
+> cd /data
+> cp -r home/shared/* shared/
+> rm -rf home/shared/
+> ```
+> And then:
+> ```shell
+> chown -R 1000:100 /data/home/*
+> chown -R 1000:100 /data/shared/*
+> ```
+> This is because all users have the same uid in QHub v0.4 onward.
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:165]:
```{suggestion}
> From QHUb v0.4. all users will have the same `uid`
```

File: docs/source/admin_guide/backup.md [hunk 40]
```diff
@@ -145,9 +147,27 @@ tar -xvf 2021-04-23.tar
...
+>
+
 ### Google cloud provider
 
-To use the Google Cloud provider, install [gsutil](https://cloud.google.com/storage/docs/gsutil_install). The instructions are same as the preceding steps. Additionally, use these commands for copy/download of the backup:
+To use the Google Cloud provider, install the [gsutil](https://cloud.google.com/storage/docs/gsutil_install) CLI. The instructions are the same as the preceding steps. Additionally, use these commands for copy/download of the backup:
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:170]: Again what are the preceding steps? do you refer to AWS? if so make it explicit otherwise folks will get lost

File: docs/source/admin_guide/backup.md [hunk 41]
```diff
@@ -160,7 +180,50 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
 ### Digital Ocean
 
 Similar instructions, but use Digital Ocean spaces. This guide explains installation of the command-line tool:
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:182]:
```{suggestion}
We suggest you follown this Digital Ocean installation guide of the command-line tool:
```

File: docs/source/admin_guide/backup.md [hunk 42]
```diff
@@ -160,7 +180,50 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
-https://www.digitalocean.com/community/tutorials/how-to-migrate-from-amazon-s3-to-digitalocean-spaces-with-rclone
+<https://www.digitalocean.com/community/tutorials/how-to-migrate-from-amazon-s3-to-digitalocean-spaces-with-rclone>
+
+
+## JupyterHub Database
+
+The JupyterHub database will mostly be recreated whenever you start a new cluster, but should be backed up to save Dashboard configurations.
+
+You want to do something very similar to the NFS backup, above - this time you need to back up just one file, and it's on the PersistentVolume `hub-db-dir`.
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:190]:
```{suggestion}
You want to do something very similar to the NFS backup, above - this time you need to back up just one file located in the  PersistentVolume `hub-db-dir`.
```

File: docs/source/admin_guide/backup.md [hunk 43]
```diff
@@ -160,7 +180,50 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+
+First, you might think you can just make a new `pod.yaml` file, this time specifying `claimName: "hub-db-dir"` instead of `claimName: "jupyterhub-dev-share"`. However, `hub-db-dir` is 'Read Write Once' - the 'Once' meaning it can only be mounted to one pod at a time, but the JupyterHub pod will already have this mounted! So the same approach will not work here.
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:192]:
```{suggestion}
It is important to note that `hub-db-dir` is 'Read Write Once' - the 'Once' meaning it can only be mounted to one pod at a time. in this case the `JupyterHub pod`. Instead of creating a new `pod.yml` file you will need to follow the next steps:

```

File: docs/source/admin_guide/backup.md [hunk 44]
```diff
@@ -160,7 +180,50 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+
+So instead of mounting to a new 'debugger pod' we have to access the JupyterHub pod directly and see what we can do from there.
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:194]:
```{suggestion}
Instead of mounting to a new 'debugger pod' you have to access the JupyterHub pod directly using the `kubectl` CLI.
```

File: docs/source/admin_guide/backup.md [hunk 45]
```diff
@@ -160,7 +180,50 @@ gsutil cp gs://<your_bucket_name>/backups/2021-04-23.tar .
...
+
+```bash
+kubectl exec -n dev --stdin --tty hub-765c9488d6-8z4nj -- /bin/bash
+```
+
+There is no need to TAR anything up since the only file required to be backed up is `/srv/jupyterhub/jupyterhub.sqlite`.
+
+### Backing up JupyterHub DB
+
+Now we just need to upload the file to S3. You might want to [install the AWS CLI tool](#installations) as we did before, but unfortunately the Hub container is quite locked down and it isn't straightforward to install that... So instead you need to upload to S3 using curl directly:
```
Comment (trallard) [inline docs/source/admin_guide/backup.md:213]:
```{suggestion}
Now we just need to upload the file to S3. You might want to [install the AWS CLI tool](#installations) as we did before, however, as the Hub container is a rather restricted environment the recommended approach is to upload files to AWS S3 buckets using curl.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 46]
```diff
@@ -0,0 +1,174 @@
...
+- Back up user data by following [this backup guide](./backup.md).
+- Change the existing cluster to a different URL (for example `qhub-old.mycompany.com`) so it is hidden.
+- Run `qhub upgrade` to make recommended modifications to the `qhub-config.yaml` file.
+- Deploy a new QHub with the desired version (to your original preferred URL for example `qhub.mycompany.com` but a new project_name to avoid resource name clashes).
+- Restore user data to the new cluster.
+- Once you have confirmed the backup and upgrade have been successfully completed, you can destroy the old cluster.
+
+Full details of the steps appear below.
+
+Please always check [release notes](../dev_guide/changelog.md) for more details - and in all cases, backup your data before upgrading.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:14]:
```{suggestion}
Please always check the [release notes](../dev_guide/changelog.md) for more details - and in all cases, backup your data before upgrading.
```

File: docs/source/admin_guide/breaking-upgrade.md [hunk 47]
```diff
@@ -0,0 +1,174 @@
...
+
+> The rest of this guide assumes you are upgrading from version v0.3.14 to v0.4.
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:16]: Is this always true or can folks upgrade from any `v0.3.x`? if so it would be best to replace with that

File: docs/source/admin_guide/breaking-upgrade.md [hunk 48]
```diff
@@ -0,0 +1,174 @@
...
+
+Perform manual backups of the NFS data and JupyterHub database (ignore the section about Keycloak data since that will not exist in your v0.3.14 cluster).
+
+For more details please refer to the [Manual Backup guide](./backup.md).
+
+## 2. Rename existing QHub URL
+
+This will allow the existing cluster to remain alive in case it is needed, but the idea would be not to have it in use from now on.
+
+In the qhub-config.yaml for example:
```
Comment (trallard) [inline docs/source/admin_guide/breaking-upgrade.md:32]:
```{suggestion}
In the `qhub-config.yaml` for example:
```

**Review by trallard — CHANGES_REQUESTED** [review]
Thanks for the ping Dan - this reads much better now. Again  left some comments and suggestions. 

Another observation the file `backup.md` still has a rather inconsistent tone: the first part reads formal referring to `the user... has to/wants` and the last few sections use `you` I like the use of `you` better as is rather informal but as it is reads odd

File: docs/source/admin_guide/breaking-upgrade.md (hunk 47 above)
Comment (danlester) [reply docs/source/admin_guide/breaking-upgrade.md:16]: Will add 'or earlier'

Comment (danlester) [reply docs/source/admin_guide/breaking-upgrade.md:16]: Will add 'or earlier'

File: docs/source/admin_guide/backup.md (hunk 37 above)
Comment (danlester) [reply docs/source/admin_guide/backup.md:70]: `dev` isn't referring to the namespace in this case, but I see the confusion so will explain.

File: docs/source/admin_guide/backup.md (hunk 43 above)
Comment (danlester) [reply docs/source/admin_guide/backup.md:192]: I appreciate we jump into 'story telling mode' here, but I think I should leave it as anyone genuinely trying to understand will need to get to the bottom of this. As you say, everyone else is just copying and pasting anyway!

File: docs/source/admin_guide/awss3curl.md (hunk 32 above)
Comment (danlester) [reply docs/source/admin_guide/awss3curl.md:3]: Good idea, and I wonder if this can be imposed automatically. I have added an issue https://github.com/Quansight/qhub/issues/1113 - I haven't taken the time to manually format today.

In general, there may be guidance that you would propose for the entire documentation rather than just this review, so maybe we should make some rules for a top-down review of everything.

**Comment by danlester** [comment]
I have also changed from "the user" to "you" in `backup.md`

**Comment by trallard** [comment]
LGMT now - not sure if anyone else needs reviewing before merging, otherwise I am marking as merge ready

**Comment by trallard** [comment]
LGMT now - not sure if anyone else needs reviewing before merging, otherwise I am marking as merge ready
